### ISSUE TITLE
Converting saved checkpoints to pb file

### ISSUE THREAD

**Issue by andrewginns** [LynnHo/CycleGAN-Tensorflow-Simple#7]
I'm trying to convert the saved cktp files into a pb file.

I'm using the following to do so:

```
import tensorflow as tf

saver = tf.train.import_meta_graph('./checkpoints/pre_norm2pro/Epoch_(209)_(1100of2100).ckpt.meta', clear_devices=True)
graph = tf.get_default_graph()
input_graph_def = graph.as_graph_def()
sess = tf.Session()
saver.restore(sess, "./checkpoints/pre_norm2pro/Epoch_(209)_(1100of2100).ckpt.data")
```

The errors I'm getting are:

`2018-06-04 16:23:27.608892: W tensorflow/core/util/tensor_slice_reader.cc:95] Could not open ./checkpoints/pre_norm2pro/Epoch_(209)_(1100of2100).ckpt.data-00000-of-00001: Data loss: not an sstable (bad magic number): perhaps your file is in a different file format and you need to use a different restore operator?
2018-06-04 16:23:27.608914: W tensorflow/core/framework/op_kernel.cc:1192] Data loss: Unable to open table file ./checkpoints/pre_norm2pro/Epoch_(209)_(1100of2100).ckpt.data-00000-of-00001: Data loss: not an sstable (bad magic number): perhaps your file is in a different file format and you need to use a different restore operator?`

The model was trained successfully and runs inference using test.py 

A complete log file is attached
[log.txt](https://github.com/LynnHo/CycleGAN-Tensorflow-PyTorch/files/2068939/log.txt)

Edit: Also I assume the names of the output nodes are 'pred' and 'h4'?

**Comment by LynnHo** [comment]
@andrewginns I am sorry I haven't got an experience for converting .ckpt to .pb. I think you can try
`saver.restore(sess, "./checkpoints/pre_norm2pro/Epoch_(209)_(1100of2100).ckpt")` without `.data`, that's how I restore the variables.

**Comment by andrewginns** [comment]
@LynnHo I figured it out. Created a program to create it from the .ckpt files. Useful for conversion to other model formats like Intel Movidius graphs.

```
import os, argparse

import tensorflow as tf

# The original freeze_graph function
# from tensorflow.python.tools.freeze_graph import freeze_graph 

dir = os.path.dirname(os.path.realpath(__file__))

def freeze_graph(model_dir, output_node_names):
    """Extract the sub graph defined by the output nodes and convert 
    all its variables into constant 

    Args:
        model_dir: the root folder containing the checkpoint state file
        output_node_names: a string, containing all the output node's names, 
                            comma separated
    """
    if not tf.gfile.Exists(model_dir):
        raise AssertionError(
            "Export directory doesn't exists. Please specify an export "
            "directory: %s" % model_dir)

    if not output_node_names:
        print("You need to supply the name of a node to --output_node_names.")
        return -1

    # We retrieve our checkpoint fullpath
    checkpoint = tf.train.get_checkpoint_state(model_dir)
    input_checkpoint = checkpoint.model_checkpoint_path
    
    # We precise the file fullname of our freezed graph
    absolute_model_dir = "/".join(input_checkpoint.split('/')[:-1])
    output_graph = absolute_model_dir + "/frozen_model.pb"

    # We clear devices to allow TensorFlow to control on which device it will load operations
    clear_devices = True

    # We start a session using a temporary fresh Graph
    with tf.Session(graph=tf.Graph()) as sess:
        # We import the meta graph in the current default Graph
        saver = tf.train.import_meta_graph(input_checkpoint + '.meta', clear_devices=clear_devices)

        # We restore the weights
        saver.restore(sess, input_checkpoint)

        # We use a built-in TF helper to export variables to constants
        output_graph_def = tf.graph_util.convert_variables_to_constants(
            sess, # The session is used to retrieve the weights
            tf.get_default_graph().as_graph_def(), # The graph_def is used to retrieve the nodes 
            output_node_names.split(",") # The output node names are used to select the usefull nodes
        ) 
		
	# fix nodes
        for node in output_graph_def.node:
          if node.op == 'RefSwitch':
            node.op = 'Switch'
            for index in xrange(len(node.input)):
              if 'moving_' in node.input[index]:
                node.input[index] = node.input[index] + '/read'
          elif node.op == 'AssignSub':
            node.op = 'Sub'
            if 'use_locking' in node.attr: del node.attr['use_locking']

        # Finally we serialize and dump the output graph to the filesystem
        with tf.gfile.GFile(output_graph, "wb") as f:
            f.write(output_graph_def.SerializeToString())
        print("%d ops in the final graph." % len(output_graph_def.node))

    return output_graph_def

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument("--model_dir", type=str, default="", help="Model folder to export")
    parser.add_argument("--output_node_names", type=str, default="", help="The name of the output nodes, comma separated.")
    args = parser.parse_args()

    freeze_graph(args.model_dir, args.output_node_names)
```

This also accounts for strange node renaming due to the use of a moving_mean. Done through the #fix nodes section

**Comment by elham1992** [comment]
hi  Mr
i use this code for convert checkpoint ( .meta, .data and .index ) model files into one graph.pb file... but  error from output_node_names .... please Explain for me about this parametr in deep neural network.
